packages/react-instantsearch-ssr-nextjs/src/NextInstantSearchSSR.tsx

```diff
@@ -0,0 +1,57 @@
+import { ServerInsertedHTMLContext } from 'next/navigation';
+import React, { useContext, useRef } from 'react';
+import {
+  InstantSearch,
+  InstantSearchRSCContext,
+  useInstantSearchContext,
+  useRSCContext,
+} from 'react-instantsearch-core';
+
+import type { UiState } from 'instantsearch.js';
+import type { ReactElement } from 'react';
+import type {
+  InstantSearchProps,
+  PromiseWithState,
+} from 'react-instantsearch-core';
+
+export type NextInstantSearchSSRProps<
+  TUiState extends UiState = UiState,
+  TRouteState = TUiState
+> = {
+  children: ReactElement;
+} & InstantSearchProps<TUiState, TRouteState>;
+
+export function NextInstantSearchSSR<
+  TUiState extends UiState = UiState,
+  TRouteState = TUiState
+>({
+  children,
+  ...instantSearchProps
+}: NextInstantSearchSSRProps<TUiState, TRouteState>) {
+  const promiseRef = useRef<PromiseWithState<void> | null>(null);
+  const insertHTML =
```

Comment:
I think we should keep `insertHTML` out of `react-instantsearch-core`, it's Next-specific and other RSC implems would probably use something different for passing data from server to client, so the RSC context should only have a ref to a promise.
What we can do is before calling `searchOnlyWithDerivedHelpers` we do `promiseRef.current.then(/* insertHtml logic */)`, but we might have to make sure that it properly inserts the `<script>` tags before serving the response.

Reply:
If it does not properly wait, have a component _before_ `children` that reassigns `promiseRef` to `promiseRef.then(/* insertHTML */)`

Reply:
Here it's in the separate `react-instantsearch-ssr-nextjs`, not in core.

Reply:
This is in `react-instantsearch-core` : https://github.com/algolia/instantsearch/pull/5830/files#diff-d5ddf6e0b8b366a56a1b63f942409045d3394790e12650e1be7ca8e0d43a6f25
As well as the logic for inserting a `<script>` tag and getting `initialResults`, this should all be in the new package as it's Next.js specific.
`react-instantsearch-core` should only bring RSC support with `use`, hydration however is a Next.js specific concern

Reply:
I think we should try to put the most possible in the new package

Reply:
Because of dependencies in both `useWidget` and `useInstantSearchApi`, we need to colocate `useRSCContext` and `InstantSearchRSCContext` in react-instantsearch-core as well, or else we'd have circular deps issues I think.

Reply:
It's fine for the context to be in `react-instantsearch-core`, it has to be there anyway.
I just think `insertHTML` can be done solely in this package though. I'll try after I'm done with my RFC :)